File: CppSolution/CountAndSay.cpp
#include <iostream>
#include <sstream>
#include <map>
using namespace std;

class Solution
{
public:
    string countAndSay1(string s)
    {
        string result;
        int count = 1;
        if (s.size() == 1)
        {
            result += '1';
            result += s[0];
            return result;
        }
        for (int i = 1; i < s.size(); i++)
        {
            if (s[i] != s[i - 1])
            {
                result += count + '0';
                result += s[i - 1];
                count = 1;
            }
            else
                count++;
            if (i == s.size() - 1)
            {
                result += count + '0';
                result += s[i];
            }
        }
        return result;
    }
    string countAndSay(int n)
    {
        string t = "1";
        if (n == 1) return t;
        for (int i = 1; i < n; i++)
        {
            t = countAndSay1(t);
        }        
        return t;
    }
};

int main()
{
    Solution s;
    s.countAndSay(5);
    return 0;
}

File: JavaSolution/src/FaceBook/CountandSay.java
package FaceBook;

/**
 * Created by xyunpeng on 1/24/16.
 */
public class CountandSay {
    public String countAndSay(int n) {
        if (n == 1) return "1";
        if (n == 2) return "11";
        String result = "11";
        for (int i = 3; i <= n; i++) {
            StringBuilder t = new StringBuilder();
            int count = 0;
            for (int j = 0; j < result.length(); j++) {
                if (j > 0 && result.charAt(j - 1) != result.charAt(j)) {
                    t.append(String.valueOf(count) + result.charAt(j - 1));
                    count = 1;
                } else {
                    count++;
                }
            }
            t.append(String.valueOf(count) + result.charAt(result.length() - 1));
            result = t.toString();
        }
        return result;
    }
}


//题目描述不是很具体,可以看这个链接:  http://www.careercup.com/question?id=4425679
//这种字符串处理的题真是烦, 没什么算法还edge case特别多